src/extensions/default/PhpTooling/main.js

```diff
@@ -0,0 +1,252 @@
+        CodeInspection = brackets.getModule("language/CodeInspection"),
+        DefaultProviders = brackets.getModule("languageTools/DefaultProviders"),
+        CodeHintsProvider = require("CodeHintsProvider").CodeHintsProvider,
+        DefaultEventHandlers = brackets.getModule("languageTools/DefaultEventHandlers"),
+        PreferencesManager  = brackets.getModule("preferences/PreferencesManager"),
+        Strings             = brackets.getModule("strings"),
+        Dialogs             = brackets.getModule("widgets/Dialogs"),
+        DefaultDialogs      = brackets.getModule("widgets/DefaultDialogs"),
+        Commands               = brackets.getModule("command/Commands"),
+        CommandManager         = brackets.getModule("command/CommandManager"),
+        StringUtils             = brackets.getModule("utils/StringUtils");
+
+    var clientFilePath = ExtensionUtils.getModulePath(module, "client.js"),
+        clientName = "PhpClient",
+        _client = null,
+        phpConfig = {
+            enablePhpTooling: true,
+            executablePath: "php",
+            memoryLimit: "4095M",
+            validateOnType: "false"
+        },
+        DEBUG_OPEN_PREFERENCES_IN_SPLIT_VIEW  = "debug.openPrefsInSplitView",
+        phpServerRunning = false,
+        serverCapabilities,
+        currentRootPath;
+
+    PreferencesManager.definePreference("php", "object", phpConfig, {
+        description: Strings.DESCRIPTION_PHP_TOOLING_CONFIGURATION
+    });
+
+    PreferencesManager.on("change", "php", function () {
+        var newPhpConfig = PreferencesManager.get("php");
+
+        if ((newPhpConfig["executablePath"] !== phpConfig["executablePath"])
+                || (newPhpConfig["enablePhpTooling"] !== phpConfig["enablePhpTooling"])) {
+            phpConfig = newPhpConfig;
+            runPhpServer();
+            return;
+        }
+        phpConfig = newPhpConfig;
+    });
+
+    var handleProjectOpen = function (event, directory) {
+        if(serverCapabilities["workspace"] && serverCapabilities["workspace"]["workspaceFolders"]) {
+            _client.notifyProjectRootsChanged({
+                foldersAdded: [directory.fullPath],
+                foldersRemoved: [currentRootPath]
+            });
+            currentRootPath = directory.fullPath;
+        } else {
+            _client.stop()
+                .done(function () {
+                    setTimeout(function () {
+                        _client.start({
+                            rootPath: directory.fullPath
+                        }).done(handlePostPhpServerStart);
+                    }, 1500);
+                }).fail(function () {
+                    console.log("Error encountered while stoping Php Server.");
+                });
+        }
+    };
+
+    function registerToolingProviders() {
+        var chProvider = new CodeHintsProvider(_client),
+            phProvider = new DefaultProviders.ParameterHintsProvider(_client),
```

Comment:
@niteskum Please override the hasParameterHints function to handle <no parameter> bug. Use `TokenUtils` to check if the previous token is of type "variable", and return true only then.

Reply:
@shubhsnov  what is the bug?